expression/builtin_time.go

```diff
@@ -3518,8 +3519,17 @@ func (c *unixTimestampFunctionClass) getFunction(ctx sessionctx.Context, args []
 		if argEvaltp == types.ETString {
 			// Treat types.ETString as unspecified decimal.
 			retDecimal = types.UnspecifiedLength
+			var tmpStr string
+			var err error
+			o := true
 			if cnst, ok := args[0].(*Constant); ok {
-				tmpStr, _, err := cnst.EvalString(ctx, chunk.Row{})
+				tmpStr, _, err = cnst.EvalString(ctx, chunk.Row{})
+			} else if sf, ok := args[0].(*ScalarFunction); ok && sf.FuncName.L == ast.GetVar {
+				tmpStr, _, err = sf.EvalString(ctx, chunk.Row{})
+			} else {
+				o = false
+			}
+			if o {
```

Comment:
How about removing `o`, and check `err` first, then check if `tmpStr` is empty before calling `LastIndex`?

Reply:
Can we make sure that temStr is not ""?

Reply:
Can we make sure that tmpStr is not ""?

Reply:
Here is the output of MySQL:
```
mysql> SELECT UNIX_TIMESTAMP('');
+--------------------+
| UNIX_TIMESTAMP('') |
+--------------------+
|           0.000000 |
+--------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+------------------------------+
| Level   | Code | Message                      |
+---------+------+------------------------------+
| Warning | 1292 | Incorrect datetime value: '' |
+---------+------+------------------------------+
1 row in set (0.00 sec)
```
so when `tmpStr` is "" then, the `retDecimal` should be kept as `types.UnspecifiedLength` and later set to `6`, to make it compatible with MySQL.

Another problem is that, we may need to consider not only `Constant` and the special `GetVar()` `ScalarFunction` here, but also other scalar functions, and `Column` / `CorrelatedColumn` as well, for example in MySQL:
```
mysql> create table t(a varchar(1024));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t values('2015-11-13 10:20:19');
Query OK, 1 row affected (0.01 sec)

mysql> select unix_timestamp(a) from t;
+-------------------+
| unix_timestamp(a) |
+-------------------+
| 1447381219.000000 |
+-------------------+
1 row in set (0.00 sec)

mysql> select * from t t1 where exists (select * from t t2 where unix_timestamp(t1.a) > 0);
+---------------------+
| a                   |
+---------------------+
| 2015-11-13 10:20:19 |
+---------------------+
1 row in set (0.00 sec)
```
we can see that `unix_timestamp` should work for `Column` and `CorrelatedColumn` also, that is to say, `args[0]` can be any kind of `Expression` implementation.

Reply:
> Another problem is that, we may need to consider not only Constant and the special GetVar() ScalarFunction here, but also other scalar functions, and Column / CorrelatedColumn as well

I can't understand what you mean here, for  Column or CorrelatedColumn, TiDB has processed them correctly. As you can see from the example, `select unix_timestamp(a) from t;` don't truncate the tail zero.
Another problem I found from your post is that `SELECT UNIX_TIMESTAMP('');` in TiDB will return `NULL`, and MySQL will return 0.000000.
And the warning is `Warning | 1292 | invalid time format: ''` , which is not same with MySQL.

Reply:
Oh you are right, we have handled `Column` and `CorrelatedColumn` properly now, sorry for the misleading, we can just focus on the other problem.